src/Compilers/Core/Portable/SourceFileResolver.cs

```diff
@@ -39,21 +48,71 @@ public SourceFileResolver(ImmutableArray<string> searchPaths, string baseDirecto
 
             _baseDirectory = baseDirectory;
             _searchPaths = searchPaths;
-        }
+            _pathMap = pathMap;
 
-        public string BaseDirectory
-        {
-            get { return _baseDirectory; }
+            // the keys in pathMap should not end with a path separator
+            if (!pathMap.IsDefaultOrEmpty)
+            {
+                foreach (var kv in pathMap)
+                {
+                    var key = kv.Key;
+                    if (key == null || key.Length == 0)
+                    {
+                        throw new ArgumentException(CodeAnalysisResources.EmptyKeyInPathMap, nameof(pathMap));
+                    }
+
+                    var value = kv.Value;
+                    if (value == null || value.Length == 0)
```

Comment:
It feels like an empty value should be OK. For example if one wants to remove root from the path.